Feature/22 쓰레기통 이미지 등록, 조회 API 구현

7zrv opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->
resolved : #22  <!-- 관련된 이슈를 적어주세요 #이슈번호 -->
## 📌 쓰레기통 이미지 등록과 조회 기능 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

## 👩‍💻 쓰레기통 등록시 이미지가 같이 등록, 조회시 이미지 url을 같이 조회

- 기존 TrashcanApiController의 생성, 조회 엔드포인트 수정
- TrashcanService에 이미지 등록, 조회 로직을 추가 했으며 TrashcanImgService를 주입받아 이미지 관련 기능을 이용하도록 했습니다.
- s3 bucket 관련 상수들을 yml 파일의 환경변수로 옮겼으며 환경변수는 시스템 실행후 값을 받아오는 방식이므로 상수가 아닌 일반 변수로 바꿔주었습니다 ex) BASE_URL -> baseUrl
- yml 파일에 cloud 관련 항목을 만들어 하위 항목에 s3 관련 설정을 추가했습니다
- yml 파일에 multipart 관련 설정을 추가했습니다. 파일 하나의 최대크기, 요청 한번당 총 크기 제한등의 설정입니다.
- 파일 업로드 관련 예외처리 코드 추가
- 쓰레기통 정보, 이미지 url을 같이 반환하는 dto 추가


## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

기존 로직에 이미지 관련 기능이 추가적으로 들어간거라 실수가 있는 부분들이 있을거 같습니다
어색한 부분이 있다면 말씀해주시면 감사하겠습니다.

게시판 이미지와 업로드 부분을 공유하려 했으나 아직 통합을 못한 상태로
쓰레기통 이미지의 경우 TrashcanImgServiceImpl에 s3로 업로드 하는 메서드들이 들어가 있습니다.
해당 부분은 추후에 분리, 통합을 하도록 하겠습니다.

s3Config의 경우는 게시판과 공유중인데 네이밍을 일부 수정했습니다.
게시판 로직에 영향이 가지 않도록 신경썻으나 실수가 있을지 모르니 한번씩 확인해주시면 감사하겠습니다.

7zrv committed:
Feat: 쓰레기통 이미지 등록, 조회 API 구현

- 기존 TrashcanApiController의 생성, 조회 엔드포인트 수정
- TrashcanService에 이미지 등록, 조회 로직을 추가 했으며
  TrashcanImgService를 주입받아 이미지 관련 기능을 이용하도록 했습니다.
- s3 bucket 관련 상수들을 yml 파일의 환경변수로 옮겼으며
  환경변수는 시스템 실행후 값을 받아오는 방식이므로 상수가 아닌 일반 변수로
  바꿔주었습니다 ex) BASE_URL -> baseUrl
- yml 파일에 cloud 관련 항목을 만들어 하위 항목에 s3 관련 설정을 추가했습니다
- yml 파일에 multipart 관련 설정을 추가했습니다. 파일 하나의 최대크기, 요청 한번당 총 크기 제한등의 설정입니다.
- 파일 업로드 관련 예외처리 코드 추가
- 쓰레기통 정보, 이미지 url을 같이 반환하는 dto 추가

jhk01007 commented on src/main/java/com/sscanner/team/global/exception/ExceptionCode.java:
```diff
@@ -47,7 +47,8 @@ public enum ExceptionCode {
     DUPLICATED_PHONE(409, "이미 가입된 핸드폰 번호입니다. "),
     PASSWORD_NOT_MATCH(400, "비밀번호와 비밀번호 확인이 일치하지 않습니다."),
-    USER_NOT_FOUND(400,"해당 유저가 존재하지 않습니다.");
+    USER_NOT_FOUND(400,"해당 유저가 존재하지 않습니다."),
```
저희 컨벤션 상 NOT_EXIST_USER가 더 적절할 것 같습니다!

7zrv replied:
죄송합니다 에러 코드 수정중에 user 관련 에러의 행이 변경된 것 같습니다!
유저 담당이신 지은님께 전달해드리겠습니다

Leehunil commented on src/main/java/com/sscanner/team/trashcan/service/TrashcanImgServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+        fileValidate.add(".JPEG");
+        fileValidate.add(".PNG");
+        String idxFileName = fileName.substring(fileName.lastIndexOf("."));
+        if (!fileValidate.contains(idxFileName)) {
```
저는 시간복잡도에서 List.contains()는 O(n)이고, HashSet.contains()는 O(1)로 알고 있습니다!! 그래서 HashSet 사용하시는게 근소한 차이긴 하지만 성능이 조금 더 좋아질 것 같습니다!!

7zrv replied:
확인하고 수정 해놓겠습니다 감사합니다!

Leehunil reviewed (approved):
그  외에는 머지해도 될 것 같습니다~

7zrv committed:
Fix: 쓰레기통 업로드시 파일 검사 메소드 수정

- 확장자 검사시 기존 List -> Hashset으로 변경하여 검사 속도 개선
- 파일명, 확장자명에 대한 예외처리 추가
- 파일명에 대한 에러코드 추가
